### ISSUE TITLE
provisionComposer.js - Missed concepts on how it should actually be functional for us

### ISSUE THREAD

**Issue by ramonck** [microsoft/BotFramework-Composer#3367]
Hi Srs.,

1) 90% of the users will want to utilize a currently running resource group and not create a new one, the script always wants to create it first, I don't think this should be the correct approach in the first place it should try to find and reutilize the resource group I'm informing and then if it doesn't exist try to create one.
2) Those 90% will probably not have permissions to create a resource group because it's a corporation and there's admins vs developers and we're on the developers side that should be the mindset here for this script the admin is going to just give permissions or copy things around but not create Bots.
3) A lot of people will not have the setup like you want which is Name-Environment some will want the hostname setup and you're forcing us to follow this path which is not good, should be optional hostname or Name-Environment
4) When I ran your script it told me that it created the AppID, I don't want it to create the AppID, I want it to fetch the AppID that exists, that's 90% of the cases. It should always be try to fetch first and then if it doesn't exist then move to create mode, this for me has been an issue for this script.

I think you should think about this as a script to help users to create additional resources and not only create everything from scratch.

**Comment by luhan2017** [comment]
@ramonck yes, you can definitely leverage your existing resources for deployment. when you try to add a new publish profile in composer, you will see the schema as following:

you can fill in your exist settings to make it work:

1. If you just want to deploy your bot to ABS without configuring any other services, you can just use a profile as the following:
```
{
  "accessToken": "<your access token>",
  "hostname": "<your web app name>",  
  "settings": {
    "MicrosoftAppId": "<the appid of your bot channel registration>",
    "MicrosoftAppPassword": "<the app password of your bot channel registration>"
  }
}
```
2. If you have luis configured in your composer bot, you should use this:
```
{
  "accessToken": "<your access token>",
  "hostname": "<your web app name>",
  "luisResource": "<your luis service name>", 
  "settings": {
    "luis": {
      "endpointKey": "<your luis endpointKey>",
      "endpoint": "<your luis prediction service endpoint, by-default is https://${luis.region}.api.cognitive.microsoft.com>"
      "authoringKey": "<your luis authoringKey>",
      "authoringEndpoint": "<your luis authoring service endpoint, by-default is https://${luis.region}.api.cognitive.microsoft.com>"
      "region": "<your luis region, for example westus>"
    },
    "MicrosoftAppId": "<the appid of your bot channel registration>",
    "MicrosoftAppPassword": "<the app password of your bot channel registration>"
  }
}

3. you can add other settings on demand. 

4. last, you can always deploy your bot in visual studio by eject the runtime and copy your bot assets to ComposerDialogs folder in the runtime/azurewebapp or runtime/azurefunctions folder.
update the appsettings.deployment.json

Please let me know if it not works.

**Comment by benbrown** [comment]
Thanks for the feedback @ramonck - this is useful input as we plan the next phase of integration with Azure.

**Comment by ramonck** [comment]
> @ramonck yes, you can definitely leverage your existing resources for deployment. when you try to add a new publish profile in composer, you will see the schema.
> 
> you can fill in your exist settings to make it work:
> 
> 1. If you just want to deploy your bot to ABS without configuring any other services, you can just use a profile as the following:
> 
> ```
> {
>   "accessToken": "<your access token>",
>   "hostname": "<your web app name>",  
>   "settings": {
>     "MicrosoftAppId": "<the appid of your bot channel registration>",
>     "MicrosoftAppPassword": "<the app password of your bot channel registration>"
>   }
> }
> ```
> 
> 1. If you have luis configured in your composer bot, you should use this:
> 
> ```
> {
>   "accessToken": "<your access token>",
>   "hostname": "<your web app name>",
>   "luisResource": "<your luis service name>", 
>   "settings": {
>     "luis": {
>       "endpointKey": "<your luis endpointKey>",
>       "endpoint": "<your luis prediction service endpoint, by-default is https://${luis.region}.api.cognitive.microsoft.com>"
>       "authoringKey": "<your luis authoringKey>",
>       "authoringEndpoint": "<your luis authoring service endpoint, by-default is https://${luis.region}.api.cognitive.microsoft.com>"
>       "region": "<your luis region, for example westus>"
>     },
>     "MicrosoftAppId": "<the appid of your bot channel registration>",
>     "MicrosoftAppPassword": "<the app password of your bot channel registration>"
>   }
> }
> ```
> 
> 1. you can add other settings on demand.
> 2. last, you can always deploy your bot in visual studio by eject the runtime and copy your bot assets to ComposerDialogs folder in the runtime/azurewebapp or runtime/azurefunctions folder.
>    update the appsettings.deployment.json
> 
> Please let me know if it not works.

@luhan2017 Thank you for this I figured some of this out looking at the other tickets but it's good as a reference to have this consolidated here, thank you.
I tried the first set of instructions and I currently only have a Azure Function in place in my Resource Group I want to use that Azure Function, what should I fill in for the hostname?
If I generate a token with the "az account get-access-token" will that be enough for accessing the Azure function?
I'm getting the following error: Access denied due to invalid subscription key or wrong API endpoint. Make sure to provide a valid key for an active subscription and use a correct regional API endpoint for your resource.
Not sure if that small set there is enough to Identify that I want it as an Azure Function and want to deploy into an existing Azure Function.

@benbrown That was the idea to help you guys understand were we are coming from and how we can make this all workout, I think the tool is great it would be cool if I could just login to azure and it would fetch my stuff and I just select it but whatever you guys can do let us go with it.

**Comment by luhan2017** [comment]
@ramonck 
For Azure Function, first you should choose "Publish bot to an Azure Functions" while adding a new publish profile. and you can just use the Azure Function name as the hostname. 

For this error: "I'm getting the following error: Access denied due to invalid subscription key or wrong API endpoint. Make sure to provide a valid key for an active subscription and use a correct regional API endpoint for your resource."

It could because you didn't set a correct hostname. This is my example of the settings and a screenshot of azure resource:

```
{
  "accessToken": "***********",
  "name": "Luhan0603AF",
  "environment": "dev",
  "hostname": "Luhan0603AF-dev",
  "luisResource": "Luhan0603AF-dev-luis",
  "settings": {
    "applicationInsights": {
      "InstrumentationKey": "**********"
    },
    "luis": {
      "endpointKey": "************",
      "authoringKey": "",
      "region": "westus"
    },
    "MicrosoftAppId": "***********",
    "MicrosoftAppPassword": "***********"
  }
}
```

![image](https://user-images.githubusercontent.com/32191031/84641251-59035080-af2d-11ea-9dfc-f8cde88471da.png)

Please let me know if it still not works.

**Comment by ramonck** [comment]
Hi @luhan2017,

Thank you for the example there, I found out what my first problem is maybe I will be able to deploy after this first one is solved.

I tried creating the bot in https://dev.botframework.com/ and then migrating it to Azure with the migrate button, the problem is that my account is restricted to a Resource Group and I don't have permissions to create Azure AAD so basically it said it was "migrated" but in the end it didn't actually migrate it just migrated from a UI perspective but the actual APP ID continued to reside on the Botframework.com side and from a deployment perspective I was not able to deploy to due lack of permissions with that App ID.

I just requested to create a new APP ID for this solution specifically and will try again after I have it in place.

Thanks
